feat: add dark theme support

xymoryn opened the pull request:
Resolve #11

<img width="960" height="915" alt="图片" src="https://github.com/user-attachments/assets/127be405-e037-43fb-8ae8-810d9475af89" />

xymoryn committed:
feat: add dark theme support for pages

xymoryn committed:
fix: prevent Monaco editor iframe size reset on theme change

xymoryn committed:
perf: add theme pre-initialization to prevent flash of unstyled content

xymoryn committed:
feat: sync Vuetify theme with page theme

xymoryn committed:
feat: add dark theme support to Monaco editor

xymoryn committed:
refactor(css): optimize theme color variable names

xymoryn committed:
refactor(css): remove unused theme color variables

xymoryn committed:
style(css): improve theme contrast ratios

xymoryn committed:
refactor(XCodeEditor): automatically sync with Vuetify theme

xymoryn commented:
Hi @ichaoX , just a friendly ping to see if this PR is on the right track.

Is there anything else needed from my end to move this forward? Thanks!

ichaoX commented on src/assets/view/options.vue:
```diff
@@ -752,6 +764,10 @@ export default {
+    this.$root.$on('theme-changed', () => {
+      this.$forceUpdate();
+    });
+
```
Is the `theme-changed` event still necessary? I noticed no difference after removing it.

xymoryn replied:
Thanks for the heads up, I have updated the PR.

The `theme-changed` event is redundant. The Vuetify theme is already updated by the [util._theme.onThemeChange](https://github.com/ichaoX/ext-webRequest/pull/13/files#diff-53665b2ca0240ef1fe22bfc3e86da6aa00c5de29bcb85c7420ef29c4378f3aebR108-R110) callback  in `main.js`.

ichaoX commented on src/assets/view/options.vue:
```diff
@@ -635,6 +635,19 @@
+                  <x-list-item label="Theme">
+                    <v-chip-group
+                      v-model="settings.theme_mode"
+                      @change="onThemeChange"
```
Use the `saveSettings` method to save the settings.

```suggestion
                      @change="saveSettings({ theme_mode: $event })"
```

ichaoX commented on src/assets/main.js:
```diff
@@ -91,8 +91,24 @@ const routes = [
+
+util.getSettings(['theme_mode'], results => {
+    util.initTheme(results.theme_mode || 'auto');
+});
```
Listen for option changes

```suggestion
util.getSettings(['theme_mode'], results => {
    util.initTheme(results.theme_mode || 'auto');
}, true);
```

ichaoX replied:
Listen for option changes in another context

```suggestion
((context0) => {
    util.getSettings(['theme_mode'], results => {
        util.initTheme(results.theme_mode || 'auto');
    }, true);
    util.context(context0);
})(util.context('page'));
```

xymoryn committed:
refactor(options): remove redundant theme-changed listener

Removes the unnecessary `theme-changed` listener from the `mounted` hook. The Vuetify theme is already updated by a `util._theme.onThemeChange` callback in `main.js`.

xymoryn committed:
refactor(XExpEditor): use computed property for theme reactivity

Replaces the event-based theme synchronization with a computed property that derives from `this.$vuetify.theme.dark`. This simplifies the component and improves reactivity.

xymoryn committed:
refactor(options): remove dead code emitting unnecessary theme-changed event

ichaoX commented on src/assets/index.html:
```diff
@@ -10,6 +10,11 @@
+	Use external scripts to avoid violating CSP policies.
+	Prevent background flickering before the theme loads
+-->
+	<script src="./theme-init.js"></script>
```
I tested and found this effect to be very limited:
* When the browser theme matches the extension theme, there is no background flicker.
* When the browser theme is opposite to the extension theme, an unavoidable background flicker occurs during page load (it actually flickers faster), and only when refreshing the page does it work better (though refreshes are generally unnecessary).

Currently, I don't recommend using `localStorage` for loading synced configurations in this PR.

---

Avoid background flicker when matching browser theme, without using JS:

```html
	<meta name="color-scheme" content="light dark" />
	<style media="(prefers-color-scheme: dark)">
		:root {
			background-color: #121212;
			color: #fff;
		}
	</style>
```

xymoryn replied:
My theme generally follows the system settings, but I have indeed seen cases where others use dark system theme while applications or websites remain in light theme.

During testing, I noticed a perceptible flicker when opening or refreshing pages if the extension theme didn't match the system theme, hence I added this fix. From my observation, the flickering is significantly reduced compared to before the fix.

But as you mentioned, this fix introduces additional complexity while being rarely needed, so I've removed it in the new PR.

I've also implemented all other optimization suggestions. Please review them again when you have time.

ichaoX commented on src/assets/util.js:
```diff
@@ -367,4 +368,63 @@ ${exp}
+    },
+    getCurrentTheme() {
+        return this._theme.resolved;
+    },
```
It is recommended to add namespaces when there are a large number of non-core methods.

```suggestion
    theme: {
        _state: {...},
        init(),
        ...
    },
```

ichaoX commented on src/assets/editor.css:
This file is actually an external library.

Please add additional CSS files instead of modifying this one.

ichaoX commented on src/assets/editor.css:
```diff
@@ -9,7 +50,7 @@
     resize: vertical;
     overflow-y: hidden;
-    background-color: #eee;
+    background-color: var(--theme-editor-bg);
```
This modification seems unnecessary as it will be overridden by vuetify's `.theme--light.v-card`.

ichaoX commented on src/assets/view/popup.vue:
```diff
@@ -247,12 +246,22 @@ export default {
-  color: gray;
+  color: var(--theme-text-muted);
 }
 </style>
```
Perhaps we can use Vuetify's CSS classes instead.

```diff
-        <v-list class="rule-list" dense two-line subheader>
+        <div v-if="!ruleStat.length" class="pa-5 text-center text--secondary">
+          No matching rules
+        </div>
+        <v-list v-else class="rule-list" dense two-line subheader>
```

ichaoX commented on src/assets/view/components/XExpEditor.vue:
```diff
@@ -42,6 +42,7 @@
             ? {
                 params: params,
                 timeout: settings.code_editor_idle_timeout,
+                theme: currentTheme,
```
Is this still necessary? It seems to still work fine after I remove it.

ichaoX commented on src/assets/view/components/XExpEditor.vue:
```diff
@@ -493,18 +497,14 @@ export default {
+    inset: 0;
+    height: 100% !important;
+    resize: none;
+    background: var(--theme-bg);
```
It is not recommended to switch to using `inset`, as its browser compatibility is slightly worse.

---

Perhaps we can use `v-card` to get the background color.

```diff
-    <div v-else @mouseleave="pinHint = false">
+    <v-card
+      v-else
+      @mouseleave="pinHint = false"
+      flat
+      tile
+      :class="{ fullscreen }"
+    >
       <component
         :is="useCodeEditor ? 'x-code-editor' : 'v-textarea'"
-        :class="{ fullscreen }"
         ref="input"
```

xymoryn committed:
feat(theme): Refactor theme initialization for performance and simplicity

ichaoX commented on src/assets/view/popup.vue:
```diff
@@ -241,18 +243,12 @@ export default {
 }
+
+/* Avoid background color differences on pages */
+
```
This can be deleted.

ichaoX commented on src/assets/main.js:
```diff
@@ -91,8 +91,27 @@ const routes = [
 
+const vuetify = new Vuetify({
+    // Initialize with pre-calculated theme to prevent flash of wrong theme
+    theme: {
+        dark: window.matchMedia && window.matchMedia('(prefers-color-scheme: dark)').matches,
+    },
```
Perhaps there is a better way to reuse this part of the logic.

```suggestion
    theme: {
        dark: util.theme._resolve() == 'dark',
    },
```

ichaoX commented on src/assets/util.js:
```diff
@@ -367,4 +368,60 @@ ${exp}
+        },
+        _apply(theme) {
+            const resolvedTheme = theme === 'dark' ? 'dark' : 'light';
+            document.documentElement.setAttribute('data-theme', resolvedTheme);
```
The `data-theme` attribute currently has minimal utility and can be removed.

ichaoX commented on src/assets/theme.css:
```diff
@@ -0,0 +1,21 @@
+html, body {
+    background-color: var(--theme-bg);
+    color: var(--theme-text);
+}
```
Is this only for when Vuetify hasn't finished loading yet?

After switching to `<meta name="color-scheme">` and removing the synchronous loading setting, this seems to no longer have any noticeable effect.

ichaoX commented on src/assets/view/components/XExpEditor.vue:
```diff
@@ -493,18 +498,16 @@ export default {
+    resize: none;
+    z-index: 1;
+    margin: 0;
+    padding: 1rem;
```
It's better to revert some unrelated changes.

```suggestion
  position: fixed;
  /* inset: 0; */
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  height: 100% !important;
  resize: none;
  z-index: 1;
  margin: 0;
  padding: 1rem;
```

ichaoX commented on src/assets/index.html:
```diff
@@ -9,7 +9,9 @@
 	<link href="/external/vuetify/2.x/vuetify.css" rel="stylesheet">
+	<link href="./theme.css" rel="stylesheet">
 	<link href="./editor.css" rel="stylesheet">
+	<meta name="color-scheme" content="light dark" />
```
For better compatibility, the following styles are also required:

```css
@media (prefers-color-scheme: dark) {
    :root {
        background-color: #121212;
        color: #fff;
    }
}
```

ichaoX commented on src/assets/view/popup.vue:
```diff
@@ -241,18 +243,12 @@ export default {
 }
+
+/* Avoid background color differences on pages */
+
```
This can be removed.

xymoryn replied:
Thanks for the review, the PR has been updated

xymoryn committed:
chore(theme): simplify implementation logic and clean up unused code

ichaoX reviewed (approved):
Awesome job! Thanks for the great work!
